### PR TITLE
EIP 4844 - blobhash opcode (part 1)

### PR DESCRIPTION
**Pull request by r0qs**
~Depends on https://github.com/ethereum/solidity/pull/14743~ (merged)
First part of https://github.com/ethereum/solidity/issues/14740

This PR implements the blobhash opcode as specified in the EIP-4844 and add tests for assembler mode. The opcode is not exposed to Solidity in this part; that will be addressed in part 2. While it's intended to be non-breaking, additional tests may be required.

### CODE REVIEW

File: test/EVMHost.h [hunk 1]
```diff
@@ -127,6 +127,8 @@ class EVMHost: public evmc::MockedHost
 	static evmc::Result resultWithGas(int64_t gas_limit, int64_t gas_required, bytes const& _data) noexcept;
 	static evmc::Result resultWithFailure() noexcept;
 
+	static const std::array<evmc::bytes32, 2> m_blob_hashes;
```
Comment (nikola-matic) [inline test/EVMHost.h]:
```suggestion
	static const std::array<evmc::bytes32, 2> m_blobHashes;
```

File: libevmasm/Instruction.h [hunk 2]
```diff
@@ -89,6 +89,7 @@ enum class Instruction: uint8_t
 	CHAINID,			///< get the config's chainid param
 	SELFBALANCE,		///< get balance of the current account
 	BASEFEE,            ///< get the block's basefee
+	BLOBHASH,           ///< get the blob versioned hash
```
Comment (cameel) [inline libevmasm/Instruction.h]:
```suggestion
	BLOBHASH = 0x49,    ///< get a blob's versioned hash
```

There can be more than one blob in a block and the instruction only gets one of them. Or maybe we could be a bit more verbose: `get a versioned hash of one of the blobs associated with the block`.

File: liblangutil/EVMVersion.h [hunk 3]
```diff
@@ -103,6 +103,7 @@ class EVMVersion:
 	bool hasBaseFee() const { return *this >= london(); }
 	bool hasPrevRandao() const { return *this >= paris(); }
 	bool hasPush0() const { return *this >= shanghai(); }
+	bool hasBlobhash() const { return *this >= cancun(); }
```
Comment (cameel) [inline liblangutil/EVMVersion.h]:
```suggestion
	bool hasBlobHash() const { return *this >= cancun(); }
```

File: test/EVMHost.h (hunk 1 above)
Comment (cameel) [inline test/EVMHost.h]: We don't generally use the `m_` prefix for static members. Also, I'd put such members last, to clearly visually separate them from all the static functions. Otherwise it's easy not to even notice that it's different.

```suggestion
	static std::array<evmc::bytes32, 2> const blobHashes;
```

File: test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol [hunk 4]
```diff
@@ -0,0 +1,12 @@
+contract C {
+    function f() pure external returns (bytes32 ret) {
+        assembly {
+            ret := blobhash()
+        }
```
Comment (cameel) [inline test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: You're missing a check against this in `libyul/AsmAnalysis.cpp` that the `blobbasefee` PR (#14755) has.

Interestingly, in that PR the check seems ineffective for some reason (https://github.com/ethereum/solidity/pull/14755#discussion_r1443316281), but it does work for `basefee`.

File: libyul/backends/evm/EVMDialect.cpp [hunk 5]
```diff
@@ -128,11 +128,22 @@ std::set<YulString> createReservedIdentifiers(langutil::EVMVersion _evmVersion)
 		return _instrName == "prevrandao" && _evmVersion < langutil::EVMVersion::paris();
 	};
 
+	// TODO remove this in 0.9.0. We allow creating functions or identifiers in Yul with the name
+	// blobhash for VMs before cancun.
+	auto blobHashException = [&](evmasm::Instruction _instr) -> bool
+	{
+		return _instr == evmasm::Instruction::BLOBHASH && _evmVersion < langutil::EVMVersion::cancun();
+	};
```
Comment (cameel) [inline libyul/backends/evm/EVMDialect.cpp:143]: Just a reminder to add this to the issue mentioned in https://github.com/ethereum/solidity/pull/14755#discussion_r1436083925.

File: test/libyul/yulInterpreterTests/blobhash.yul [hunk 6]
```diff
@@ -0,0 +1,13 @@
+{
+  sstore(0, blobhash(0))
+  sstore(1, blobhash(1))
+  sstore(2, blobhash(2)) // should still 0
```
Comment (cameel) [inline test/libyul/yulInterpreterTests/blobhash.yul]: I assume you meant that it should "store 0"? You should say why - I guess it's because `EVMHost` puts only two blob hashes in the block the transaction is being executed in?

File: test/tools/yulInterpreter/Interpreter.h [hunk 7]
```diff
@@ -111,6 +111,11 @@ struct InterpreterState
 	/// Number of the current state instance, used for recursion protection
 	size_t numInstance = 0;
 
+	// KZG version hash
+	const util::FixedHash<1> KZGHashVersion = util::FixedHash<1>(1);
+	// KZG commitments
+	const std::array<u256, 2> KZGCommitments = {0x01, 0x02};
```
Comment (cameel) [inline test/tools/yulInterpreter/Interpreter.h]: KZG is just a type of commitment. Technically we could even get another KZG commitment for something else in a block in the future so I would not use it in the name. The fact that the commitments are for blobs is more relevant for disambiguation:

```suggestion
	// Blob commitment hash version
	util::FixedHash<1> const blobHashVersion = util::FixedHash<1>(1);
	// Blob commitments
	std::array<u256, 2> const blobCommitments = {0x01, 0x02};
```

File: test/EVMHost.h (hunk 1 above)
Comment (cameel) [reply test/EVMHost.h]: Also, I see you use this only once, in the constructor, to initialize the context. Why not just put the hard-coded values there along with all the other hard-coded constants?

File: test/libsolidity/semanticTests/inlineAssembly/blobhash_empty.sol [hunk 8]
```diff
@@ -0,0 +1,11 @@
+contract C {
+    function f() public view returns (bytes32 ret) {
+        assembly {
+            ret := blobhash(3)
```
Comment (cameel) [inline test/libsolidity/semanticTests/inlineAssembly/blobhash_empty.sol]: I'd add a comment that this being empty depends on the fact that we have 2 blobs hard-coded for the current block in `EVMHost`. Out of that context is not at all clear why you'd be guaranteed a 0 here.

Also changing the name to something like `blobhash_index_exceeding_blob_count.sol` would also add more clues.

**Review by cameel — COMMENTED** [review]
This is missing a changelog entry (we usually add a separate one for the opcode and for high-level support in Solidity). Also, the bits of docs from #14759 describing the opcode should really be in this PR.

Other than that I did not find any serious problems. Mostly small things, easy to tweak. I added a fixup to deal with some of it, the rest is in comments.

Since the evmone update is now merged, I rebased the PR on `develop`.

File: test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol (hunk 4 above)
Comment (nikola-matic) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: This is incredibly odd - when EVMDialect is created - more specifically the builtins (`m_functions`), `basefee` is not in there - but then when it check for the presence of `basefee` (for Berlin), it does end up being there, which is why it also throws the 'only supported in London` error.

Comment (nikola-matic) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: Ah, OK, I *think* I know why - going through the debugger when running a test (e.g. if you run a test under a specific EVM version (in my case Berlin)), it seems that two dialects are created - the one you specified, and the the current default - in this case, Berlin and Shanghai; Berlin obviously doesn't have `basefee`, but Shanghai does, and is thus added to `m_functions`, which is why it's found as a builtin, and which is why it throws the "only supported in XXX EVM version" error.

`blobbasefee` is obviously in neither Berlin, nor Shanghai, so this check is skipped, since it's not found in the builtins, and then just throws the generic (function not defined) error at the end.

My inkling is that once we bump the default soltest/isoltest version to Cancun, that this behaviour will equalize for all Cancun and pre-Cancun identifiers.

Comment (nikola-matic) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: Again, I may be mistaken, since I can't see this mechanism anywhere in the code...

Comment (cameel) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: If that's the cause, should be easy to verify by locally changing the default in `EVMVersion` and rerunning the test.

File: libevmasm/Instruction.h (hunk 2 above)
Comment (r0qs) [reply libevmasm/Instruction.h]: I believe it's better to state that it is associated with the transaction, right? Because the `index` in `blobhash(index)` represents the index of the blob within the transaction. Although EIP-4844 doesn't provide details on this, upon reviewing it and other associated EIPs like [EIP-2718: Typed Transaction Envelope](https://eips.ethereum.org/EIPS/eip-2718), it appears to be the case. Initially, I thought it was related to a beacon block number, but it doesn't seem to be the case

File: test/libyul/yulInterpreterTests/blobhash.yul (hunk 6 above)
Comment (r0qs) [reply test/libyul/yulInterpreterTests/blobhash.yul]: Exactly. I added a comment there.

File: test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol (hunk 4 above)
Comment (nikola-matic) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]:
```
    TypeError 1006: (74-91): "blobbasefee" is not supported by the VM version.
    TypeError 6679: (187-198): The "blobbasefee" instruction is only available for Cancun-compatible VMs (you are currently compiling for "berlin").
    DeclarationError 8678: (180-200): Variable count for assignment to "ret" does not match number of values (1 vs. 0)
```
Confirmed.

File: libyul/backends/evm/EVMDialect.cpp (hunk 5 above)
Comment (r0qs) [reply libyul/backends/evm/EVMDialect.cpp:143]: The removal of it was mentioned here: https://github.com/ethereum/solidity/issues/14770

File: test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol (hunk 4 above)
Comment (r0qs) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: Hum...so should we add a test for it anyway?

Comment (cameel) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: Nice. In that case the test is fine as is.

Comment (cameel) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: Actually, I'd still add the check in `AsmAnalysis.cpp` even if we cannot test it properly yet. The current test won't fail when we make it the default and we'll very likely forget then.

File: libevmasm/Instruction.h (hunk 2 above)
Comment (cameel) [reply libevmasm/Instruction.h]: Right, my bad. It's per transaction, not per block.

File: test/EVMHost.h (hunk 1 above)
Comment (r0qs) [reply test/EVMHost.h]: Indeed, I changed it to what you suggested. My initial thought was that it could be used in the `EVMInstructionInterpreter` as well, but it turned out that it doesn't.

File: libevmasm/Instruction.h [hunk 9]
```diff
@@ -89,6 +89,7 @@ enum class Instruction: uint8_t
 	CHAINID,			///< get the config's chainid param
 	SELFBALANCE,		///< get balance of the current account
 	BASEFEE,            ///< get the block's basefee
+	BLOBHASH = 0x49,    ///< get a versioned hash of one of the blobs associated with the transaction
```
Comment (nikola-matic) [inline libevmasm/Instruction.h:92]: You don't need an explicit value for this - `BASEFEE` is `0x48`. Important to note - if `blobbasefee` ends up getting merged first, it would be nice if you could get rid of the `0x4a` assignment when rebasing.

Comment (cameel) [reply libevmasm/Instruction.h:92]: We don't, but I think it's better to have the explicit value. Less error prone than counting the values and easier to see the value when you need it (e.g. in review).

File: docs/yul.rst [hunk 10]
```diff
@@ -925,6 +925,8 @@ the ``dup`` and ``swap`` instructions as well as ``jump`` instructions, labels a
 +-------------------------+-----+---+-----------------------------------------------------------------+
 | blockhash(b)            |     | F | hash of block nr b - only for last 256 blocks excluding current |
 +-------------------------+-----+---+-----------------------------------------------------------------+
+| blobhash(i)             |     | N | versioned blob hash of blob index i                             |
```
Comment (cameel) [inline docs/yul.rst]:
```suggestion
| blobhash(i)             |     | N | versioned hash of transaction's i-th blob               |
```

File: Changelog.md [hunk 11]
```diff
@@ -1,7 +1,7 @@
 ### 0.8.24 (unreleased)
 
 Language Features:
-
+ * Yul: Introduce builtin ``blobhash()`` for retrieving a versioned blob hash of a given blob index within the transaction.
```
Comment (cameel) [inline Changelog.md]:
```suggestion
 * Yul: Introduce builtin ``blobhash()`` for retrieving a versioned blob hash of a given blob index within the transaction.

```

File: libevmasm/Instruction.h (hunk 9 above)
Comment (nikola-matic) [reply libevmasm/Instruction.h:92]: I agree, but then we should do so for all of them; obviously not in this PR. I suggested he remove it for the sake of consistency.

File: test/EVMHost.cpp [hunk 12]
```diff
@@ -137,6 +137,14 @@ EVMHost::EVMHost(langutil::EVMVersion _evmVersion, evmc::VM& _vm):
 	// The minimum value of basefee
 	tx_context.block_base_fee = evmc::bytes32{7};
 
+	static evmc_bytes32 const blob_hashes_array[] = {
+		0x0100000000000000000000000000000000000000000000000000000000000001_bytes32,
+		0x0100000000000000000000000000000000000000000000000000000000000002_bytes32
+	};
+
+	tx_context.blob_hashes = blob_hashes_array;
+	tx_context.blob_hashes_count = sizeof(blob_hashes_array) / sizeof(blob_hashes_array[0]);
```
Comment (cameel) [inline test/EVMHost.cpp]: Can we simplify it even more?
```suggestion
	tx_context.blob_hashes = {
		0x0100000000000000000000000000000000000000000000000000000000000001_bytes32,
		0x0100000000000000000000000000000000000000000000000000000000000002_bytes32
	};
	tx_context.blob_hashes_count = 2;
```

Comment (r0qs) [reply test/EVMHost.cpp]: Oh, actually not, because `tx_context.blob_hashes` is a pointer to `evmc_bytes32`. But the count can be hardcoded for sure.

File: libevmasm/Instruction.h (hunk 9 above)
Comment (cameel) [reply libevmasm/Instruction.h:92]: I don't think we'll ever do this properly so I'd rather leave existing entries as is and just add values to new or modified ones as we change them over time.

File: test/EVMHost.cpp (hunk 12 above)
Comment (cameel) [reply test/EVMHost.cpp]: Well, if we have to have the separate variable, then having the count set automatically based on it was actually better :) Now that I think of it, I probably could have used `tx_context.blob_hashes()` in my suggestion too.

File: Changelog.md [hunk 13]
```diff
@@ -1,6 +1,7 @@
 ### 0.8.24 (unreleased)
 
 Language Features:
+ * Yul: Introduce builtin ``blobhash()`` for retrieving a versioned blob hash of a given blob index within the transaction.
```
Comment (cameel) [inline Changelog.md]: How about this wording?
```suggestion
 * Yul: Introduce builtin ``blobhash()`` for retrieving versioned hashes of blobs associated with the transaction.
```

File: test/libsolidity/syntaxTests/viewPureChecker/blobhash_not_pure.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/viewPureChecker/blobhash_not_pure.sol:1]: We should also add `blobhash` to the `syntaxTests/viewPureChecker/inline_assembly_instructions_allowed...` group of tests

**Review by cameel — COMMENTED** [review]
I looked through all the uses of the already existing `blockhash()` to make sure we're not forgetting anything we should have for `blobhash()` here and found only one minor test that might need it.

~Other than that, we actually should change https://github.com/ethereum/solidity/pull/14757#discussion_r1449024031 back (I'll post details in a moment).~ EDIT: Actually it's fine. I wanted to note that the variable assigned to the pointer will go out of scope, but it's static so it's fine.

I'm going to look through #14759 before approving this in case something from there should still end up in this PR, but if not, we'll be done here.

File: test/EVMHost.cpp (hunk 12 above)
Comment (cameel) [reply test/EVMHost.cpp]: Actually, looking at it again, I see why you wanted to put it in the class initially. I was assuming `tx_context.blob_hashes` was holding a copy and thought you made the variable `static` only as a micro-optimization.

If it's a pointer and we have to hold on to the memory then yeah, the original would be fine. But I guess hiding it as a static variable in a function is fine too.

File: test/libyul/yulSyntaxTests/blobhash_pre_cancun.sol [hunk 14]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (matheusaaguiar) [inline test/libyul/yulSyntaxTests/blobhash_pre_cancun.sol]: I guess there is no problem this being a `.sol` file ? However, all the other tests are `.yul`, so maybe we should keep the pattern.

Comment (r0qs) [reply test/libyul/yulSyntaxTests/blobhash_pre_cancun.sol]: Oh, indeed. Fixed ;)

File: test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol (hunk 4 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: > Again, I may be mistaken, since I can't see this mechanism anywhere in the code...

I think I figured out why it was done this way: https://github.com/ethereum/solidity/pull/14737#discussion_r1450767916. IMO it should really have had a comment explaining the intent (or at least hint at it through better naming).

In any case, another conclusion is that we could just make this error an assert for now to avoid having to ignore it via `error-codes.py` and only make it an error when we make `cancun` the default (the assert will helpfully fail then).

Comment (r0qs) [reply test/libsolidity/syntaxTests/inlineAssembly/blobhash_pre_cancun.sol:5]: Right. I made it an assertion. See https://github.com/ethereum/solidity/pull/14757/commits/d186d1cb237e1440ffe9f409f2d9ef07ec993ed0 and if it is ok I will squash the commits ;)

File: docs/grammar/SolidityLexer.g4 [hunk 15]
```diff
@@ -305,7 +305,7 @@ YulEVMBuiltin:
 	| 'delegatecall' | 'staticcall' | 'return' | 'revert' | 'selfdestruct' | 'invalid'
 	| 'log0' | 'log1' | 'log2' | 'log3' | 'log4' | 'chainid' | 'origin' | 'gasprice'
 	| 'blockhash' | 'coinbase' | 'timestamp' | 'number' | 'difficulty' | 'prevrandao'
-	| 'gaslimit' | 'basefee' | 'blobbasefee';
+	| 'gaslimit' | 'basefee' | 'blobhash' | 'blobbasefee';
```
Comment (cameel) [inline docs/grammar/SolidityLexer.g4]: These are not sorted alphabetically so I'd put it next to `blockhash`.

Same in other places where the order is arbitrary, e.g. `Instructions.cpp`.

Comment (cameel) [reply docs/grammar/SolidityLexer.g4]: Oh, I see that it's between `basefee` and `blobbasefee` in the opcode numbering. That's probably why you ordered it like this.

I'd still put it next to `blockhash` where possible though, because they're similar in many ways and it makes reviewing easier when they're next to each other.

File: libyul/AsmAnalysis.cpp [hunk 16]
```diff
@@ -705,6 +706,9 @@ bool AsmAnalyzer::validateInstructions(evmasm::Instruction _instr, SourceLocatio
 		);
 	};
 
+
+	// The errors below are only issued when the undeclared identifier is a builtin on the default EVM version
+	// but not on the currently selected one, since `validateInstructions` uses the default EVM version.
```
Comment (cameel) [inline libyul/AsmAnalysis.cpp]: Nice. I'd just clarify it a bit since there are two functions called `validateInstructions()` and this one does not actually use the default EVM version.

```suggestion
	// The errors below are meant to be issued when processing an undeclared identifier matching a builtin name present on the default EVM version
	// but not on the currently selected one, since the other `validateInstructions()` overload uses the default EVM version.
```

File: docs/grammar/SolidityLexer.g4 (hunk 15 above)
Comment (r0qs) [reply docs/grammar/SolidityLexer.g4]: Yes, I tried to follow the opcode numbering order. Ok, I will move it close to `blockhash` then.

File: libyul/AsmAnalysis.cpp [hunk 17]
```diff
@@ -729,6 +733,9 @@ bool AsmAnalyzer::validateInstructions(evmasm::Instruction _instr, SourceLocatio
 		errorForVM(5430_error, "only available for London-compatible");
 	else if (_instr == evmasm::Instruction::BLOBBASEFEE && !m_evmVersion.hasBlobBaseFee())
 		errorForVM(6679_error, "only available for Cancun-compatible");
+	else if (_instr == evmasm::Instruction::BLOBHASH)
+		// TODO: Upgrade this assertion to an error, similar to the ones above, when Cancun becomes the default EVM version.
```
Comment (nikola-matic) [inline libyul/AsmAnalysis.cpp]: Why is this a warning and why doesn't it have a `cancun` check (i.e. `hasBlobHash`)? Should I do the same for `blobbasefee` above? If this should remain as is, I'll add this to the `0.9.0` issue, because we could easily forget about this.

**Review by nikola-matic — COMMENTED** [review]
Aside from the one question I left (which is important, at least for me), looks good.

File: libyul/AsmAnalysis.cpp (hunk 17 above)
Comment (cameel) [reply libyul/AsmAnalysis.cpp]: It's not a warning, it's an assertion that we'll never run into this in the current code. We won't forget about it - that's the advantage of it being an assert - it will fail in the PR where we change the default EVM version and then we'll know we have to make it an error. IMO this arrangement is slightly better than having a dead error that we can't really trigger and test. It's also a bit clearer (makes it obvious that this can't be triggered) and we also won't forget to remove the error code from the whitelist. Still, it's not *that* much of a difference so I wouldn't bother changing it in your PR. I'd have suggested it there but I only realized we could do this after you merged it. This will be irrelevant anyway once we change the default.

> why doesn't it have a `cancun` check (i.e. `hasBlobHash`)

That's actually a good question. I mean, in practice it does not matter since the only thing it changes is the return value from `validateInstructions()` in case where the opcode exists, but that's [the one case where we ignore that return value](https://github.com/ethereum/solidity/blob/eabe457bf94cd785e0b097b4c03e54572101d29a/libyul/AsmAnalysis.cpp#L327C3-L327C23).

The way the checks are done for other opcodes above, we jump to `else` and return `false` when they do exist, i.e. when the validation is successful, which seems wrong. @r0qs's check is correct (it will return `true` instead), but also inconsistent with that behavior. I guess we should fix that by making them consistent and not ignoring the result (in a separate PR), but it's also just a very low priority cosmetic thing since it won't really affect current behavior in any way.

Comment (nikola-matic) [reply libyul/AsmAnalysis.cpp]: I don't see how having an assert here is better than having a test pinned to `shanghai` that will also fail when we upgrade to `cancun` - compare [blobbasefee test](https://github.com/ethereum/solidity/pull/14755/files#diff-173b4adaa431f4cfc08788f4d23ba19ad95d255388294bda53c9b0d7a6204bd8) with the [basefee test](https://github.com/ethereum/solidity/pull/14755/files#diff-77a7a77f2a2831156ab9f1652dabc5af0858e8ff6739dd3890540ff1d3c5f875); the former will fail due to unmet expectations (dead error becoming live) when we upgrade, and I'd prefer to review test expectation updates rather than code changes when we do upgrade.

Comment (cameel) [reply libyul/AsmAnalysis.cpp]: I think it's better because it it's clearer. This part of code definitely needs more clarity - it already required too much effort to figure out what's happening there and why. This case is impossible to hit right now and the assert expresses exactly that.

Comment (r0qs) [reply libyul/AsmAnalysis.cpp]: > The way the checks are done for other opcodes above, we jump to `else` and return `false` when they do exist, i.e. when the validation is successful, which seems wrong. @r0qs's check is correct (it will return `true` instead), but also inconsistent with that behavior. I guess we should fix that by making them consistent and not ignoring the result (in a separate PR), but it's also just a very low priority cosmetic thing since it won't really affect current behavior in any way.

Indeed. I will move that condition to a separated `if` outside the that chain of checks, so maybe it is a bit more clear of what is going on. But yeah, the idea is exactly that, that it should never reach this point if the instruction is `BLOBHASH` and the current EVM version does not support it.

File: libyul/AsmAnalysis.cpp [hunk 18]
```diff
@@ -705,6 +706,16 @@ bool AsmAnalyzer::validateInstructions(evmasm::Instruction _instr, SourceLocatio
 		);
 	};
 
+	// The errors below are meant to be issued when processing an undeclared identifier matching a builtin name
+	// present on the default EVM version but not on the currently selected one,
+	// since the other `validateInstructions()` overload uses the default EVM version.
+	if (_instr == evmasm::Instruction::BLOBHASH)
+	{
+		// TODO: Upgrade this assertion to an error, similar to the ones above, when Cancun becomes the default EVM version.
```
Comment (matheusaaguiar) [inline libyul/AsmAnalysis.cpp]:
```suggestion
		// TODO: Upgrade this assertion to an error, similar to the ones below, when Cancun becomes the default EVM version.
```
Annoying nitpick, but I could not find anything else :D

Comment (r0qs) [reply libyul/AsmAnalysis.cpp]: Ops! Thanks ;)

File: libyul/AsmAnalysis.cpp (hunk 17 above)
Comment (cameel) [reply libyul/AsmAnalysis.cpp]: I mean, I think it was better (and less verbose) at the original spot. I'd just do `yulAssert(false)` instead of reporting an error. But well, not like it matters that much.

One thing I think you should adjust though is to put the comment explaining the errors actually above the errors, not above the assert. Otherwise it's likely that it will be removed by mistake along with the assert when we switch the default.

Comment (r0qs) [reply libyul/AsmAnalysis.cpp]: Oh right. Sorry, my bad, I got a bit confused by the discussion. I moved it back because I agree that the initial change was better. So hopefully, we can get this merged ;)

**Review by cameel — APPROVED** [review]

File: Changelog.md [hunk 19]
```diff
@@ -3,7 +3,7 @@
 Language Features:
  * Introduce global ``block.blobbasefee`` for retrieving the blob base fee of the current block.
  * Yul: Introduce builtin ``blobbasefee()`` for retrieving the blob base fee of the current block.
-
+ * Yul: Introduce builtin ``blobhash()`` for retrieving versioned hashes of blobs associated with the transaction.
```
Comment (cameel) [inline Changelog.md:6]: Oh, one more thing, let's finally stop messing with the spacing in the changelog :)

```suggestion
 * Yul: Introduce builtin ``blobhash()`` for retrieving versioned hashes of blobs associated with the transaction.
 
```

I fixed it earlier in all those PRs but it keeps reappearing again and again for some reason. Don't worry about the wrong spacing of the `Compiler Features:` section. We'll fix it when we sort the changelog. In PRs just add an entry for your PR and don't touch other lines.
